consensus/reactor.go

```diff
@@ -1,550 +1,535 @@
-	conR.Logger.Info("Reactor ", "waitSync", conR.WaitSync())
+// NewReactor returns a reference to a new consensus reactor, which implements
+// the service.Service interface. It accepts a logger, consensus state, references
+// to relevant p2p Channels and a channel to listen for peer updates on. The
+// reactor will close all p2p Channels when stopping.
+func NewReactor(
+	logger log.Logger,
+	cs *State,
+	stateCh *p2p.Channel,
+	dataCh *p2p.Channel,
+	voteCh *p2p.Channel,
+	voteSetBitsCh *p2p.Channel,
+	peerUpdates *p2p.PeerUpdatesCh,
+	waitSync bool,
+	options ...ReactorOption,
+) *Reactor {
+
+	r := &Reactor{
+		conS:          cs,
+		waitSync:      waitSync,
+		peers:         make(map[p2p.NodeID]*PeerState),
+		Metrics:       NopMetrics(),
+		stateCh:       stateCh,
+		dataCh:        dataCh,
+		voteCh:        voteCh,
+		voteSetBitsCh: voteSetBitsCh,
+		peerUpdates:   peerUpdates,
+		closeCh:       make(chan struct{}),
+	}
+	r.BaseService = *service.NewBaseService(logger, "Consensus", r)
+
+	for _, opt := range options {
+		opt(r)
+	}
+
+	return r
+}
+
+// OnStart starts separate go routines for each p2p Channel and listens for
+// envelopes on each. In addition, it also listens for peer updates and handles
+// messages on that p2p channel accordingly. The caller must be sure to execute
+// OnStop to ensure the outbound p2p Channels are closed.
+func (r *Reactor) OnStart() error {
+	r.Logger.Debug("consensus wait sync", "wait_sync", r.WaitSync())
 
 	// start routine that computes peer statistics for evaluating peer quality
-	go conR.peerStatsRoutine()
+	//
+	// TODO: Evaluate if we need this to be synchronized via WaitGroup as to not
+	// leak the goroutine when stopping the reactor.
+	go r.peerStatsRoutine()
 
-	conR.subscribeToBroadcastEvents()
+	r.subscribeToBroadcastEvents()
 
-	if !conR.WaitSync() {
-		err := conR.conS.Start()
-		if err != nil {
+	if !r.WaitSync() {
+		if err := r.conS.Start(); err != nil {
 			return err
 		}
 	}
 
+	go r.processStateCh()
+	go r.processDataCh()
+	go r.processVoteCh()
+	go r.processVoteSetBitsCh()
+	go r.processPeerUpdates()
+
 	return nil
 }
 
-// OnStop implements BaseService by unsubscribing from events and stopping
+// OnStop stops the reactor by signaling to all spawned goroutines to exit and
+// blocking until they all exit, as well as unsubscribing from events and stopping
 // state.
-func (conR *Reactor) OnStop() {
-	conR.unsubscribeFromBroadcastEvents()
-	if err := conR.conS.Stop(); err != nil {
-		conR.Logger.Error("Error stopping consensus state", "err", err)
+func (r *Reactor) OnStop() {
+	r.unsubscribeFromBroadcastEvents()
+
+	if err := r.conS.Stop(); err != nil {
+		r.Logger.Error("failed to stop consensus state", "err", err)
+	}
+
+	if !r.WaitSync() {
+		r.conS.Wait()
 	}
-	if !conR.WaitSync() {
-		conR.conS.Wait()
+
+	// wait for all spawned peer goroutines to gracefully exit
+	r.mtx.Lock()
+	for _, ps := range r.peers {
+		ps.closer.Close()
+		ps.broadcastWG.Wait()
```

Comment:
I don't think we should wait while holding the mutex. Maybe just grab a copy of the map while holding the mutex, then iterate once to close, then iterate again to wait (so they close concurrently).